pkg/health/checker/checker.go

```diff
@@ -256,18 +393,169 @@ func publishLatestHealth(inCh <-chan api.KVPairs, quitCh <-chan struct{}, result
 }
 
-func nodeIDsToStatusEndpoints(nodeIds []types.NodeName, status manifest.StatusStanza) []string {
-	statusEndpoints := make([]string, len(nodeIds))
+func nodeIDToStatusEndpoint(nodeID types.NodeName, status manifest.StatusStanza) string {
 	scheme := "https"
 	if status.HTTP {
 		scheme = "http"
 	}
-	for i, nodeId := range nodeIds {
-		statusEndpoints[i] = fmt.Sprintf("%s://%s:%d%s", scheme, nodeId, status.Port, status.GetPath())
+	return fmt.Sprintf("%s://%s:%d%s", scheme, nodeID, status.Port, status.GetPath())
+}
+
+func nodeIDsToStatusEndpoints(nodeIDs []types.NodeName, status manifest.StatusStanza) []string {
+	statusEndpoints := make([]string, len(nodeIDs))
+
+	for i, nodeID := range nodeIDs {
+		statusEndpoints[i] = nodeIDToStatusEndpoint(nodeID, status)
 	}
 	return statusEndpoints
 }
 
+// retrieves replication controller managed nodeIDs for a serviceID
+func (h shadowTrafficHealthChecker) serviceNodesRC(serviceID string) ([]types.NodeName, error) {
+	rcIDs, err := h.resourceClient.RCIDsForPod(types.PodID(serviceID))
+	if err != nil {
+		return nil, err
+	}
+
+	var nodeIDs []types.NodeName
+	for _, rcID := range rcIDs {
+		rc, err := h.rcStore.Get(rcID)
+		if err != nil {
+			return nil, err
+		}
+		// get hostnames for rc
+		labeled, err := h.labelReader.GetMatches(rc.NodeSelector, labels.NODE)
+		if err != nil {
+			return nil, err
+		}
+		// extract nodes
+		for _, node := range labeled {
+			nodeIDs = append(nodeIDs, types.NodeName(node.ID))
+		}
+	}
+	return nodeIDs, nil
+}
+
+// retrieves daemon set nodeIDs for a serviceID
+func (h shadowTrafficHealthChecker) serviceNodesDS(serviceID string) ([]types.NodeName, error) {
+	dsList, err := h.resourceClient.DaemonSetsForPod(types.PodID(serviceID))
+	if err != nil {
+		return nil, err
+	}
+	var nodeIDs []types.NodeName
+	for _, ds := range dsList {
+		selector := klabels.Everything().Add(DSIDLabel, klabels.EqualsOperator, []string{ds.ID.String()})
+		labeled, err := h.labelReader.GetMatches(selector, labels.NODE)
+		if err != nil {
+			return nil, err
+		}
+		for _, node := range labeled {
+			nodeIDs = append(nodeIDs, types.NodeName(node.ID))
+		}
+	}
+	return nodeIDs, nil
+}
+
+func (h shadowTrafficHealthChecker) serviceNodes(serviceID string, checkRCNodes, checkDSNodes bool) ([]types.NodeName, error) {
+	nodeIDs := []types.NodeName{}
+	if checkRCNodes {
+		rcNodeIDs, err := h.serviceNodesRC(serviceID)
+		if err != nil {
+			return nil, err
+		}
+		nodeIDs = append(nodeIDs, rcNodeIDs...)
+	}
+	if checkDSNodes {
+		dsNodeIDs, err := h.serviceNodesDS(serviceID)
+		if err != nil {
+			return nil, err
+		}
+		nodeIDs = append(nodeIDs, dsNodeIDs...)
+	}
+	return nodeIDs, nil
+}
+
+func (h shadowTrafficHealthChecker) WatchService(
+	ctx context.Context,
+	serviceID string,
+	resultCh chan<- map[types.NodeName]health.Result,
+	errCh chan<- error,
+	watchDelay time.Duration,
+	useHealthService bool,
+	useOnlyHealthService bool,
+	status manifest.StatusStanza,
+	checkRCNodes bool,
```

Comment:
What is this for? Is the client expected to know whether they are checking an RC or a DS? Why not make them separate functions that share a helper?

Reply:
I think only the client would know whether to check RC, DS or both nodes when calling WatchService or Service. This could be a parameter when initializing the checker.